src/KetherNFTPublisher.sol

```diff
@@ -0,0 +1,231 @@
+     */
+    function setApprovalForAll(address operator, bool approved) public {
+        require(operator != _msgSender(), "ERC721: approve to caller");
+
+        _operatorApprovals[_msgSender()][operator] = approved;
+        emit ApprovalForAll(_msgSender(), operator, approved);
+    }
+
+    /**
+     * @dev See {IERC721-isApprovedForAll}.
+     */
+    function isApprovedForAll(address owner, address operator) public view returns (bool) {
+        return _operatorApprovals[owner][operator];
+    }
+
+    /**
+     * @dev Returns whether `publisher` is allowed to manage `tokenId`.
+     */
+    function isApprovedToPublish(address publisher, uint256 tokenId) public view returns (bool) {
+        address owner = ketherNFT.ownerOf(tokenId); // Implicitly checks tokenId validity
+        address approved = ketherNFT.getApproved(tokenId);
+
+        // Is this contract approved to control publishers?
+        if (!ketherNFT.isApprovedForAll(owner, address(this)) &&
+            approved != address(this)
+        ) {
+            revert(Errors.MustApprovePublisher);
+        }
+
+        // Is the publisher the owner or approved?
+        if (owner == publisher ||
+            approved == publisher ||
+            ketherNFT.isApprovedForAll(owner, publisher)
+        ) {
+            return true;
+        }
+
+        // Is Magistrate allowed to publish?
+        address magistrate = ketherSortition.getMagistrate();
+        if (magistrate != publisher) {
+            // Publisher is not the magistrate, nevermind.
+            return false;
+        }
+
+        // Sortition contract approved means the magistrate is approved
+        return (
+            getApproved(tokenId) == address(ketherSortition) ||
+            isApprovedForAll(owner, address(ketherSortition))
+        );
+    }
+
+    /**
+     * @notice Publish to a Kether ad unit as an approved publisher.
+     *
+     * Must be called by the token owner or an approved address.
+     *
+     * @dev See {KetherNFT-publish}.
+     */
+    function publish(uint _idx, string calldata _link, string calldata _image, string calldata _title, bool _NSFW) external {
+        if (!isApprovedToPublish(_msgSender(), _idx)) {
+            revert(Errors.SenderNotApproved);
+        }
+
+        // Check timeout
+        if (publishTimeout > 0) {
+            if (_tokenToPublishTimestamp[_idx] + publishTimeout > block.timestamp) {
+                revert(Errors.MustTimeout);
+            }
+            _tokenToPublishTimestamp[_idx] = block.timestamp;
+        }
+
+        // Check fee
+        if (publishFeeAmount > 0 && address(publishFeeToken) != address(0)) {
+            address tokenOwner = ketherNFT.ownerOf(_idx);
+            publishFeeToken.transferFrom(_msgSender(), tokenOwner, publishFeeAmount);
+        }
+
+        IKetherNFTPublish(address(ketherNFT)).publish(_idx, _link, _image, _title, _NSFW);
+    }
+
+
+    /*** Magistrate-managed Functions: ***/
```

Comment:
Snuck in some fresh magistrate-owned mechanisms here, will allow for nice composability with some future hypothetical ERC20 😐